src/test/java/com/rackspace/ceres/app/web/QueryControllerTest.java

```diff
@@ -0,0 +1,111 @@
+package com.rackspace.ceres.app.web;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.when;
+
+import com.rackspace.ceres.app.model.QueryResult;
+import com.rackspace.ceres.app.services.QueryService;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import reactor.core.publisher.Flux;
+
+@ActiveProfiles("test")
+@WebFluxTest(QueryController.class)
+public class QueryControllerTest {
+
+  @MockBean
+  QueryService queryService;
+
+  @Autowired
+  private WebTestClient webTestClient;
+
+  @Test
+  public void testQueryApi() {
+
+    Map<String, String> queryTags = new HashMap<>();
+    queryTags.put("os", "linux");
+    queryTags.put("deployment", "dev");
+    queryTags.put("host", "h-1");
+
+    Map<Instant, Double> values = new HashMap<>();
+    values.put(Instant.now(), 111.0);
+
+    List<QueryResult> queryResults = new ArrayList<>();
+    queryResults.add(new QueryResult().setMetricName("cpu-idle").setTags(queryTags).setTenant("t-1")
+        .setValues(values));
+
+    when(queryService.queryRaw(anyString(), anyString(), any(), any(), any()))
+        .thenReturn(Flux.just(queryResults.get(0)));
+
+    webTestClient.get()
+        .uri(uriBuilder -> uriBuilder.path("/api/query")
+            .queryParam("tenant", "t-1")
+            .queryParam("metricName", "cpu-idle")
+            .queryParam("tag", "os=linux")
+            .queryParam("start", "1d-ago")
+            .build())
+        .exchange().expectStatus().isOk()
+        .expectBodyList(QueryResult.class).isEqualTo(queryResults);
+  }
```

Comment:
Please add a verify of `queryService.queryRaw`

Reply:
I was thinking about adding this one but `startTime` and `endTime` in arguments will differ.

Reply:
Would using absolute time values in the RFC3339 help?

Reply:
Didn't get you on this one

Reply:
OK, maybe I don't understand what you meant by

> but startTime and endTime in arguments will differ

With relative-time syntax they're hard to test since the value differs by the clock time. However, if you pass something like "2020-10-23T16:00:00Z" then you can assert that fixed time value.

Reply:
In second test case I have used timestamp as start and end time in request param and added verify call there.